popde/adaptive_kde.py

```diff
@@ -224,6 +224,149 @@ def optimize_parameters(self, method='loo_cv', fom_plot_name=None):
+    def loo_cv_score(self, rescale_factors_alpha):
+        """
+        Compute the Leave-One-Out Cross-Validation (LOO-CV) score
+          based on log-likelihood.
+
+        Args:
+            rescale_factors_alpha (array-like): Array of rescale factors per
+              dimension and alpha value.
+
+        Returns:
+            float: Negative sum of log-likelihood scores for LOO-CV.
+        """
+        rescale_factors = rescale_factors_alpha[:-1]
+        alpha_choice = rescale_factors_alpha[-1]
+        from sklearn.model_selection import LeaveOneOut
+        loo = LeaveOneOut() 
+        fom = 0.
+        for train_index, test_index in loo.split(self.kde_data):
+            train_data, test_data = self.kde_data[train_index], self.kde_data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factors,
+                                  bandwidth=self.bandwidth, alpha=alpha_choice)
+            fom += np.log(awkde.evaluate(test_data))
+        return -fom
+
+    def kfold_cv_score(self, rescale_factors_alpha, seed=42):
+        """
+        Perform k-fold cross-validation
+        """
+        rescale_factors = rescale_factors_alpha[:-1]
+        alpha_choice = rescale_factors_alpha[-1]
+        from sklearn.model_selection import KFold
+        kf = KFold(n_splits=self.n_splits, shuffle=True, random_state=seed)
+        fom = []
+        for train_index, test_index in kf.split(self.kde_data):
+            train_data, test_data = self.kde_data[train_index], self.kde_data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factors,
+                                  bandwidth=self.bandwidth, alpha=alpha_choice)
+            log_kde_eval = np.log(awkde.evaluate(test_data))
+            fom.append(log_kde_eval.sum())
+        return -sum(fom)
+
+    def optimize_rescale_parameters(self, initial_rescale_factor, initial_alpha=0.0, method='kfold_cv', bounds=None):
+
+        """
+        given initial choice of rescale factors in each dimension
+        and alpha perform optimization based on 
+        cross validations on data
+        """
+        #make alpha bounds to be fixed in [0, 1]
+        if bounds is None:
+            set_bounds = ((None, None),) * len(initial_rescale_factor) + ((0, 1),)
+        else:
+            set_bounds = bounds
+        initial_choices = np.concatenate((initial_rescale_factor, initial_alpha))
+        from scipy.optimize import minimize
+        best_params = {'rescale_per_dim': None}
+        # Perform Nelder-mead based Optimization
+        if method == 'kfold_cv':
+            result = minimize(
+                    self.kfold_cv_score,        # func to minimize why negative?
```

Comment:
I don't understand the 'why negative?' comment. Either make it understandable or remove please

Reply:
removed